source/server/connection_handler_impl.cc

```diff
@@ -54,6 +65,29 @@ void ConnectionHandlerImpl::removeListeners(uint64_t listener_tag) {
   }
 }
 
+void ConnectionHandlerImpl::removeFilterChains(
+    uint64_t listener_tag, const std::list<const Network::FilterChain*>& filter_chains,
+    std::function<void()> completion) {
+  // TODO(lambdai): Merge the optimistic path and the pessimistic path.
+  for (auto& listener : listeners_) {
+    // Optimistic path: The listener tag provided by arg is not stale.
+    if (listener.second.listener_->listenerTag() == listener_tag) {
+      ASSERT(listener.second.tcp_listener_.has_value());
+      listener.second.tcp_listener_->get().removeFilterChains(filter_chains);
+      Event::DeferredTaskUtil::deferredRun(dispatcher_, std::move(completion));
```

Comment:
I asked this question in a previous round, but can you please add more comments on why we need to use the deferred run here and below. Also, again I'm wondering if this can be handled by the callback itself vs. having this logic here when it's unclear if it's actually needed based on what the callback does.

Reply:
Sure. I will add the comments.
> if this can be handled by the callback itself 
For this callback(aka completion), it's very hard to rewrite the completion. 

I have the strong preference to address the problem with `deferredRun` instead of the content of `completion`:
If you look at the two statements this line and 1 line above:
`listener.second.tcp_listener_->get().removeFilterChains(filter_chains)` is an **implicit defer delete**; however, the completion is an **immediate** invoke if I don't introduce `deferredRun`. What the callback does is not the core issue. The contract of "when to invoke `complete`" is the core.

Suppose I modify the completion, making it as wise as "hey, as the master thread, I know the `completion` callback could be invoked at worker thread earlier than I requested, so I would use these mechanisms to make sure envoy would survive". 


What do you  think if I rename 
`listener.second.tcp_listener_->get().removeFilterChains(filter_chains);` to 
`listener.second.tcp_listener_->get().deferRemoveFilterChains(filter_chains);`

Reply:
Yes I think that rename would be very helpful + a comment. Thank you!

Reply:
Renamed to deferRemoveFilterChains. That makes more sense when looking the statements here